### ISSUE TITLE
Consolidate Streams into single hierarchy and/or shared traits

### ISSUE THREAD

**Issue by daniels220** [guillep/pharo-core#13651]
**Describe the problem**
I want to add an extension method to all things that look like a WriteStream. It depends only on other public methods of WriteStream, so the implementation will be the same everywhere. But I can't actually just add it once—I have to add it to WriteStream, ZnEncodedWriteStream, ZnBufferedWriteStream, etc. For ReadStream it's even worse as some write streams are readable as well, but don't always share a superclass with even their most-similar *ReadStream cousin.

On a related note, the available protocol is inconsistent between all these various Stream classes, because different ones have had different levels of effort put into giving them a rich protocol. Having one place to put things would make this less likely as well.

**Classes involved**
Basically all classes with "Stream" in the name somewhere—most important for me being the Zinc streams which are used for files, but I'm sure others would have the same issue with e.g. ZLib, sockets, etc.

**Proposal**
Some single place I can put the method—I think ultimately it would be a cleaner implementation at least some of the time to move these disparate Stream classes into the actual Stream hierarchy, but not all the time, so almost certainly the best place for extension methods like this is a trait. Unfortunately the name of those traits really wants to be `ReadStream` and `WriteStream` (and a super-trait `PositionableStream` as well), but this clashes. Not sure what to suggest.

**Version information:**
 - Any Pharo version

**Expected development cost**
Adding a trait without migrating any existing code seems like it would be pretty easy. I bet there's a tremendous amount of duplication already, though, as well as methods exclusive to a particular Stream class but that _should_ be available elsewhere, so there's probably a fair amount of work to clean that up. I might be willing to make a quick pass through but don't have time to dive into it.

**Comment by jecisc** [comment]
Hi,

One problem of the Trait is that it is not supported in Pharo's core. The minimal code to run the Kernel includes some streams and it is not possible to add traits to those since the traits are not yet loaded.

**Comment by daniels220** [comment]
Oooh, good point. Could the trait be added to those classes later in the boot process, or would that leave Iceberg unable to consider the image "clean"/no changes? If that could work, I would look at which parts of the stream API are needed in the kernel—perhaps some of the richer API could be moved? The core Stream API is going to have differing implementations for every stream class anyway, it's the stuff built on top of it that is often duplicated.

If the Trait is a non-starter, perhaps at least some of the most important/common streams (I'm looking at you, Zinc) could be moved into the Stream hierarchy? Or at least given a hierarchy of their own—even just having a ZnStream superclass in common with all of them would _reduce_ code duplication substantially.

**Comment by svenvc** [comment]
I understand your point, I think, but the problem is larger than that.

From many points of view, the current hierarchy of system stream classes is ridiculous and silly, there are too many methods there, too many responsibilities are mixed.

A stream in its most general sense can contain any object, it is the streaming reading or writing that make it special. Being able to jump around in a stream (in the assumption that you have everything in memory in a collection) is not a fundamental property (and cannot be done in many streaming cases and is in conflict with the concept of streaming).

Why is PositionableStream a superclass then ? Why does PositionableStream implement message like (both !) #int32 and #int32: ? Why do abstract Stream classes deal with encoding, while that only makes sense for byte or character streams ? Do we need to separate reading and writing side, or not ? What about buffering ? And so on.

One reason for this situation is history. 

Another reason is that single inheritance cannot model these requirements.

So yes, traits might help. Any object that speaks (implements) a certain set of message can be used in a role where that is all that is required/expected.

But the next challenge is to define those traits.

Almost any piece of code uses streams, which leaves very little room for changes.

The newer streams use composition to separate concerns.

Since no good hierarchy or traits existed, they sometimes reimplement stuff.

Indeed, may Zn streams inherit from object, because everything else would not make sense (too much cruft being inherited, expectations from users). Maybe a shared abstract ZnReadStream and ZnWriteStream could reduce a bit of duplication, but that remains an implementation detail.

Over time, as the Zinc streams became more used in Pharo, their interfaces were polluted as well, because people kept asking for certain method that they were used to. Change is hard.

I am sure that there exist papers that cover this area.

Sven

**Comment by svenvc** [comment]
BTW, I am curious about what you want to add ?

**Comment by daniels220** [comment]
Yeah, it's a complicated issue for sure. I guess I'll just have to live with the duplication for now, but it seems worth at least having this issue as a reminder.

As far as the sorts of things I'm adding—just little quality-of-life conveniences. Some examples:

- `wrap: '()' around: <either a string, likely the result of a message send, or a block>`. The first argument can also be a Character, e.g. `$'`, `$"`.
- `spaces: 4`
- `expand: '<1p>: <2p>' with: arg1 with: arg2`, analogous to `#print:` -> `#printOn:` but for `expandMacrosWith:`. I think I had to hack core to extract an `expandMacrosOn:[with:...]` first.
- `peekForAll: 'foo'`, as `#peekFor:` but for multiple elements (consumes all if found, leaves position alone otherwise—`'foo' readStream peekForAll: 'for'; position "=> 1"`).
- `nextWhile: [:el | "consumes as long as the block answers true, positions before the element that got a false"]`, also returns the collection of elements which got a true.

There's a rich world of API possibilities for Streams, to a similar extent as with FP-inspired collection protocol (`select:`, `collect:`, et al), and it's just so easy to add them...or it would be, if I could do it in just one place...

**Comment by Ducasse** [comment]
What I suggest is: 
- one feature at a time
- covered by tests
